kernel/sched.c

```diff
@@ -1382,6 +1389,22 @@ static void halt_thread(struct k_thread *thread, uint8_t new_state)
 #ifdef CONFIG_THREAD_ABORT_NEED_CLEANUP
 		k_thread_abort_cleanup(thread);
 #endif /* CONFIG_THREAD_ABORT_NEED_CLEANUP */
+
+		/* Do this "set _current to dummy" step last so that
+		 * subsystems above can rely on _current being
+		 * unchanged.  Disabled for posix as that arch
+		 * continues to use the _current pointer in its swap
+		 * code.
+		 */
+		if (dummify && !IS_ENABLED(CONFIG_ARCH_POSIX)) {
+			z_dummy_thread_init(&_thread_dummy);
```

Comment:
Just to be sure - is it expected that we can set _current on multiple cores to the same _thread_dummy at the same time?

Reply:
It's true empirically, yeah.  The only use of that thread pointer is write-only: it's the context save code on interrupt exit.  Your intuition was the same as mine, and originally I had this set up with a per-cpu dummy.  But @npitre points out that involves significant RAM costs, and there's no reason to pay that now if we don't have a failure case.

Obviously if future arch layers do want to play weird tricks with _current, by e.g. wanting to read back stuff they wrote, even though the thread won't/can't ever be scheduled, that's doable by just checking the _THREAD_DUMMY bit in state.

Reply:
I am very curious why we need to init dummy thread here.

Reply:
(sorry, wrote this outside the conversation, pasting here to fix) The name of that routine is unfortunate. The only initialization it does is to set the flag bit; it's used to set the _current thread pointer to the dummy, which so far has only happened at CPU initialization time (so that the context switch code doesn't need a special path to enter the "first" thread).